pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -756,14 +756,58 @@ func (s *FlinkStateMachine) handleSubmittingJob(ctx context.Context, app *v1beta
 		return statusUnchanged, errors.Errorf("Could not find job %s", s.flinkController.GetLatestJobID(ctx, app))
 	}
 
-	// wait until all vertices have been scheduled and started
-	allVerticesStarted := true
-	for _, v := range job.Vertices {
-		allVerticesStarted = allVerticesStarted && (v.StartTime > 0)
+	jobStartTime := getJobStartTimeInUTC(job.StartTime)
+	now := time.Now().UTC()
+	cfg := config.GetConfig()
+	logger.Info(ctx, "Job vertex timeout config is ", cfg.FlinkJobVertexTimeout)
+	flinkJobVertexTimeout := cfg.FlinkJobVertexTimeout
+	if now.Before(jobStartTime.Add(flinkJobVertexTimeout.Duration)) {
+		allVerticesRunning, hasFailure, failedVertexIndex := monitorAllVerticesState(job)
+
+		// fail fast
+		if hasFailure {
+			s.flinkController.LogEvent(ctx, app, corev1.EventTypeWarning, "JobRunningFailed",
+				fmt.Sprintf(
+					"Vertex %d with name [%s] state is Failed", failedVertexIndex, job.Vertices[failedVertexIndex].Name))
+			return s.deployFailed(app)
+		}
+		return updateJobAndReturn(ctx, job, s, allVerticesRunning, app, hash)
+	}
+
+	s.flinkController.LogEvent(ctx, app, corev1.EventTypeWarning, "JobRunningFailed",
+		fmt.Sprintf(
+			"Not all vertice of the Flink job state is Running before timeout %f minutes", cfg.FlinkJobVertexTimeout.Minutes()))
+	return s.deployFailed(app)
+
+}
+
+func monitorAllVerticesState(job *client.FlinkJobOverview) (bool, bool, int) {
+	allVerticesRunning := true
+	// wait until all vertices have been scheduled and running
+	hasFailure := false
+	failedVertexIndex := -1
+	for index, v := range job.Vertices {
+		if v.Status == client.Failed {
+			failedVertexIndex = index
+			hasFailure = true
+			break
+		}
+		allVerticesRunning = allVerticesRunning && (v.StartTime > 0) && v.Status == client.Running
```

Comment:
If the first vertex is failing then won't this return `allVerticesRunning` as `true` when it should be `false`?

Reply:
If the vertex is failing, `hasFailure` is set to true, then line 768 if block will have early return to fail the deployment.

Reply:
Agree. Initially this part is in the main method, so I didn't set value for `allVerticesRunning` in fail scenario. I will provide the fix later